artifact/src/version.rs

```diff
@@ -0,0 +1,243 @@
+    }
+
+    /// Creates a new `ArtifactVersion` at compile time, panicking if it is
+    /// invalid.
+    pub const fn new_const(s: &'static str) -> Self {
+        match validate_version(s) {
+            Ok(()) => Self(Cow::Borrowed(s)),
+            Err(err) => panic!("{}", err.as_static_str()),
+        }
+    }
+
+    /// Returns the version as a string.
+    pub fn as_str(&self) -> &str {
+        self.0.as_ref()
+    }
+
+    /// Consumes self, returning the version as a string.
+    pub fn into_inner(self) -> Cow<'static, str> {
+        self.0
+    }
+}
+
+impl FromStr for ArtifactVersion {
+    type Err = ArtifactVersionError;
+
+    #[inline]
+    fn from_str(version: &str) -> Result<Self, Self::Err> {
+        Self::new(version)
+    }
+}
+
+impl fmt::Display for ArtifactVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl<'de> Deserialize<'de> for ArtifactVersion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let version = String::deserialize(deserializer)?;
+        validate_version(&version).map_err(serde::de::Error::custom)?;
+
+        Ok(Self(Cow::Owned(version)))
+    }
+}
+
+impl Serialize for ArtifactVersion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(self.as_str())
+    }
+}
+
+const fn validate_version(version: &str) -> Result<(), ArtifactVersionError> {
+    let len = version.len();
+
+    if len == 0 {
+        return Err(ArtifactVersionError::Empty);
+    } else if len > ArtifactVersion::MAX_LEN {
+        return Err(ArtifactVersionError::TooLong { len });
+    }
+
+    // Check that the version string matches the regex.
+    let mut b = version.as_bytes();
+    while let [first, rest @ ..] = b {
+        if !first.is_ascii_alphanumeric()
+            && !matches!(first, b'.' | b'_' | b'+' | b'-')
+        {
+            return Err(ArtifactVersionError::InvalidCharacter);
+        }
+        b = rest;
+    }
+
+    Ok(())
+}
+
+// Proptest wants regexes without anchors, so drop the first and last
+// character.
+#[cfg(any(test, feature = "proptest"))]
+static PROPTEST_REGEX: &str = {
+    let regex = ArtifactVersion::REGEX.as_bytes();
+    let [_, mid @ .., _] = regex else { unreachable!() };
+    let Ok(r) = std::str::from_utf8(mid) else { unreachable!() };
+    r
+};
+
+/// An error that occurred while creating an `ArtifactVersion`.
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+#[non_exhaustive]
+pub enum ArtifactVersionError {
+    #[error("version is empty")]
+    Empty,
+    #[error(
+        "version is too long ({len} bytes, max {})",
+        ArtifactVersion::MAX_LEN
+    )]
+    TooLong { len: usize },
+    #[error(
+        "version contains invalid character (allowed: {})",
+        ArtifactVersion::REGEX
+    )]
+    InvalidCharacter,
```

Comment:
May as well include the character that failed to parse here?

```suggestion
    #[error(
        "version contains invalid character {c:?} (allowed: {})",
        ArtifactVersion::REGEX
    )]
    InvalidCharacter { c: char },
```

Reply:
Had to use `u8` since it could be a non-ASCII byte, but good suggestion! (I learned about https://doc.rust-lang.org/std/primitive.u8.html#method.escape_ascii through this.)